astropy/wcs/wcsapi/tests/test_utils.py

```diff
@@ -56,3 +67,279 @@ def test_wcs_info_str():
+    wcs_out.world_axis_object_classes = wcs_in.world_axis_object_classes
+
+    matrix = pixel_to_pixel_correlation_matrix(wcs_in, wcs_out)
+
+    matrix = matrix.astype(int)
+
+    # The shape should be (n_pixel_out, n_pixel_in)
+    assert matrix.shape == (4, 3)
+
+    expected = np.array([[1, 1, 0], [1, 1, 0], [1, 1, 0], [0, 0, 1]])
+    assert_equal(matrix, expected)
+
+
+def test_split_matrix():
+
+    assert split_matrix(np.array([[1]])) == [([0], [0])]
+
+    assert split_matrix(np.array([[1, 1],
+                                  [1, 1]])) == [([0, 1], [0, 1])]
+
+    assert split_matrix(np.array([[1, 1, 0],
+                                  [1, 1, 0],
+                                  [0, 0, 1]])) == [([0, 1], [0, 1]), ([2], [2])]
+
+    assert split_matrix(np.array([[0, 1, 0],
+                                  [1, 0, 0],
+                                  [0, 0, 1]])) == [([0], [1]), ([1], [0]), ([2], [2])]
+
+    assert split_matrix(np.array([[0, 1, 1],
+                                  [1, 0, 0],
+                                  [1, 0, 1]])) == [([0, 1, 2], [0, 1, 2])]
+
+
+def test_efficient_pixel_to_pixel():
+
+    wcs_in = WCS(naxis=3)
+    wcs_in.wcs.ctype = 'DEC--TAN', 'FREQ', 'RA---TAN'
+    wcs_in.wcs.set()
+
+    wcs_out = WCS(naxis=3)
+    wcs_out.wcs.ctype = 'GLON-CAR', 'GLAT-CAR', 'FREQ'
+    wcs_out.wcs.set()
+
+    # First try with scalars
+    x, y, z = pixel_to_pixel(wcs_in, wcs_out, 1, 2, 3)
+    assert x.shape == ()
+    assert y.shape == ()
+    assert z.shape == ()
+
+    # Now try with broadcasted arrays
+    x = np.linspace(10, 20, 10)
+    y = np.linspace(10, 20, 20)
+    z = np.linspace(10, 20, 30)
+    Z1, Y1, X1 = np.meshgrid(z, y, x, indexing='ij', copy=False)
+    X2, Y2, Z2 = pixel_to_pixel(wcs_in, wcs_out, X1, Y1, Z1)
+
+    # The final arrays should have the correct shape
+    assert X2.shape == (30, 20, 10)
+    assert Y2.shape == (30, 20, 10)
+    assert Z2.shape == (30, 20, 10)
+
+    # But behind the scenes should also be broadcasted
+    assert unbroadcast(X2).shape == (30, 1, 10)
+    assert unbroadcast(Y2).shape == (30, 1, 10)
+    assert unbroadcast(Z2).shape == (20, 1)
+
+    # We can put the values back through the function to ensure round-tripping
+    X3, Y3, Z3 = pixel_to_pixel(wcs_out, wcs_in, X2, Y2, Z2)
+
+    # The final arrays should have the correct shape
+    assert X2.shape == (30, 20, 10)
+    assert Y2.shape == (30, 20, 10)
+    assert Z2.shape == (30, 20, 10)
+
+    # But behind the scenes should also be broadcasted
+    assert unbroadcast(X3).shape == (30, 1, 10)
+    assert unbroadcast(Y3).shape == (20, 1)
+    assert unbroadcast(Z3).shape == (30, 1, 10)
+
+    # And these arrays should match the input
+    assert_allclose(X1, X3)
+    assert_allclose(Y1, Y3)
+    assert_allclose(Z1, Z3)
+
+
+def test_efficient_pixel_to_pixel_simple():
+
+    # Simple test to make sure that when WCS only returns one world coordinate
```

Comment:
Perhaps I misunderstand this comment. Did you mean one world axis in world_out? Because the test uses 2 axes and the case with one world axis should be tested.

Reply:
Good catch, I think there was a test missing, so I've added one for the actual 1D case now (this test above was actually about correlated axes)